cmd/sftp-server-driver.go

```diff
@@ -201,19 +201,83 @@ func (f *sftpDriver) Fileread(r *sftp.Request) (ra io.ReaderAt, err error) {
 	return obj, nil
 }
 
+func (w *writerAt) Close() error {
+	for i := range w.buffers {
+		w.buffers[i] = nil
+	}
+	err := w.w.Close()
+	w.wg.Wait()
+	return err
+}
+
+const bufferLength = 1000
+
 type writerAt struct {
 	w  *io.PipeWriter
 	wg *sync.WaitGroup
+
+	nextOffset int64
+	remaining  int
+	offsets    [bufferLength]int64
+	buffers    [bufferLength][]byte
+	m          sync.Mutex
 }
 
-func (w *writerAt) Close() error {
-	err := w.w.Close()
-	w.wg.Wait()
-	return err
+func (w *writerAt) placeBytesInBuffer(b []byte, offset int64) (err error) {
+	for i := 0; i < bufferLength; i++ {
+		if len(w.buffers[i]) == 0 || w.offsets[i] == 0 {
+			w.offsets[i] = offset
+			w.buffers[i] = make([]byte, len(b))
+			copy(w.buffers[i], b)
+			w.remaining++
+			return nil
+		}
+	}
+	return errors.New("sftp segment queue is full")
+}
+
+func (w *writerAt) flushBuffers() (err error) {
+	for i := 0; i < bufferLength; i++ {
+		if w.offsets[i] == w.nextOffset {
+			n, err := w.w.Write(w.buffers[i])
+			if err != nil {
+				return err
+			}
+			if n != len(w.buffers[i]) {
+				return fmt.Errorf("expected write size %d but wrote %d bytes", len(w.buffers[i]), n)
+			}
+			w.nextOffset += int64(len(w.buffers[i]))
+			w.offsets[i] = 0
+			w.buffers[i] = nil
+			w.remaining--
+			continue
```

Comment:
You can have unflushed buffer at a lower index, that matches new offset. Either restart the loop or recurse.

If not, there is a risk you don't end flushing everything.

Reply:
We actually ran into that problem during our first round of testing.

There are two options here, one is to keep iterating the buffer until we have no offsets left that fit the order OR we can trigger a flush anytime there is anything left in the buffer. Both will result in the full file be written. 

I opted to implement it this way because I felt like it was simpler to just flush anytime there was something in the buffer. Also, there is less time between accepting packets from the SFTP clients that way.. kind of like "1 packet in, 1x flush iteration".

Reply:
Now that I think about it.. my way of doing it is probably going to scale worse then trying to evict as many as possible on every flush. 

I'll have another look at this.

Reply:
you're actually right and I can't believe I missed that.. I'll fix it asap